<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Nice)</span></div>

## 1 There is more…

### 1.1 There is more to list comprehension

In [7]:
[[a,b] for a in range(5) for b in ["A","B","C"]]

[[0, 'A'],
 [0, 'B'],
 [0, 'C'],
 [1, 'A'],
 [1, 'B'],
 [1, 'C'],
 [2, 'A'],
 [2, 'B'],
 [2, 'C'],
 [3, 'A'],
 [3, 'B'],
 [3, 'C'],
 [4, 'A'],
 [4, 'B'],
 [4, 'C']]

In [10]:
[[a,b] for a in range(10) for b in ["A","B","C"] if a%2 != 1]

[[0, 'A'],
 [0, 'B'],
 [0, 'C'],
 [2, 'A'],
 [2, 'B'],
 [2, 'C'],
 [4, 'A'],
 [4, 'B'],
 [4, 'C'],
 [6, 'A'],
 [6, 'B'],
 [6, 'C'],
 [8, 'A'],
 [8, 'B'],
 [8, 'C']]

In [11]:
nested_list=[[1, 2, 3], [4, 5, 6, 7]]
[y for x in nested_list for y in x]

[1, 2, 3, 4, 5, 6, 7]

### 1.2 Zipping a dictionary

In [12]:
fruits = ["Apple","Orange","Banana","Pear","Blueberry"]
colours = ["Red","Orange","Yellow","Green","Blue"]

dict(zip(fruits,colours))

{'Apple': 'Red',
 'Orange': 'Orange',
 'Banana': 'Yellow',
 'Pear': 'Green',
 'Blueberry': 'Blue'}

### 1.3 for and while has an else

In [44]:
numbers=[-1,-2,3]

for i in numbers:
    if i < 0: break
else:
    print('No negative numbers in the list')

## Explore 1 :  Changing a list

In [88]:
fruits = ["apple", "banana", "jackfruit",
          "pineapple", "papaya", "watermelons",
          "peaches", "durian",  "mangoes",
          "strawberries", "passionfruit"
          ]

for fruit in fruits[:]:
    if fruit[0].lower() == "p":
        fruits.remove(fruit)

print(fruits)

['apple', 'banana', 'jackfruit', 'watermelons', 'durian', 'mangoes', 'strawberries']


## Explore 2 :  A list of power

In [50]:
maximum_n = 5
result = [] * maximum_n

for n in range(1, maximum_n + 1):
    result.append([])
    result[n - 1].append(n)
    result[n - 1].append(n**2)
    result[n - 1].append(n**3)

    
print(result)

[[1, 1, 1], [2, 4, 8], [3, 9, 27], [4, 16, 64], [5, 25, 125]]


## Explore 3 :  Time profiling

In [6]:
%%timeit
list = []

for i in range(1,5):
    list.append(i**2)
    
list

1.1 µs ± 2.77 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [3]:
%%timeit
list = []
i = 1

while i < 5:
    list.append(i**2)
    i += 1  

list

996 ns ± 9.06 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [2]:
%%timeit
[i**2 for i in range(1,5)]

list

1.12 µs ± 4.51 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [5]:
%%timeit
list = []
list.append(1**2)
list.append(2**2)
list.append(3**2)
list.append(4**2)
list.append(5**2)

list

189 ns ± 0.945 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)


In [31]:
%%timeit
x = []

x += [1**2]
x += [2**2]
x += [3**2]
x += [4**2]
x += [5**2]

x

210 ns ± 2.51 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [21]:
%%timeit
arr = np.array([])

arr = np.append(arr,1**2)
arr = np.append(arr,2**2)
arr = np.append(arr,3**2)
arr = np.append(arr,4**2)
arr = np.append(arr,5**2)

arr

17.4 µs ± 79.2 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [25]:
%%timeit
arr = np.array([1,2,3,4,5])**2

arr

979 ns ± 33.1 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


|#|Option 1|Option 2|Result|
|:--|:--|:--|:--|
|1|Creating a list of squares with `for` loop|Creating a list of squares with `while` loop| Option 1: 1.1 µs ± 2.77 ns <br> Option 2: 996 ns ± 9.06 ns|
|2|Creating a list of squares with `for` loop|Creating a list of squares with a list comprehension loop.| Option 1: 1.1 µs ± 2.77 ns <br> Option 2: 1.12 µs ± 4.51 ns|
|3|Creating a list of squares using list `append()`|Creating a list of squares using list `+=`| Option 1: 189 ns ± 0.945 ns <br> Option 2: 210 ns ± 2.51 ns|
|4|Creating a list of squares using list `append()`|Creating a list of squares using `append()` of Numpy| Option 1: 189 ns ± 0.945 ns <br> Option 2: 17.4 µs ± 79.2 ns|
|5|Creating a list of squares using Numpy|Creating a list of squares using List comprehension loop.| Option 1: 979 ns ± 33.1 ns <br> Option 2: 1.12 µs ± 4.51 ns|